modules/Bio/EnsEMBL/Compara/RunnableDB/GeneTrees/OrthoTree.pm

```diff
@@ -566,5 +602,115 @@ sub check_homology_consistency {
     $self->throw("Inconsistent homologies: $bad_key") if defined $bad_key;
 }
 
+sub _create_flatfile {
+    my $self = shift;
+
+    my $outfile = $self->param('output_flatfile');
+
+    # create directory
+    my $outdir = dirname($outfile);
+    $self->run_command("mkdir -p $outdir");
+
+    # open file handle and print header
+    open( my $outfh, '>', $outfile ) or die "Cannot open $outfile for writing";
+    print $outfh join("\t", @{ $self->param_required('file_header') }) . "\n";
+    return $outfh;
+}
+
+sub _hc_flatfile {
+    my $self = shift;
+
+    my $file = $self->param('output_flatfile');
+
+    # first, check the integrity of the file
+    $self->_check_file_integrity($file);
+
+    # then HC the homologies themselves
+    print "running HCs on homologies:\n" if $self->debug;
+    my $gene_member_adaptor = $self->compara_dba->get_GeneMemberAdaptor;
+
+    open( my $fh, '<', $file ) or die "Cannot open $file for reading";
+    my $header_line = <$fh>;
+    my @header_cols = split(/\s+/, $header_line);
+
+    my $c = 1;
+    my %seen_homologies;
+    while ( my $line = <$fh> ) {
+        my $row = map_row_to_header($line, \@header_cols);
+        my ( $st_id, $hom_st_id ) = ($row->{stable_id}, $row->{homology_stable_id});
+
+        # Each homology must be linked to exactly 2 members
+        unless (defined $st_id && defined $hom_st_id) {
```

Comment:
At this stage, we have checked that all the lines in the file have the same number of columns (which  `map_row_to_header` does too). So by construction all the hash values of `$row` will be defined. The only way of getting an undefined value in `$st_id` is by `stable_id` not being a valid hash key / column header (same for `homology_stable_id` and all the `defined` below from line 663).
It seems convoluted. You can check that directly in the header line, but in fact you can check that in `fetch_input`

Reply:
Yes, this is true. TBH, I just ported all the checks from the original SQL healthchecks without fully considering whether the check really needed to be performed. I will remove this one. You are right - it would be picked up in one of the other checks.